Restores ability to eager load manually

fxn opened the pull request:
This is a draft just for the sake of discussing with concrete code.

Rails 6 does not provide an encapsulated way to eager load applications, because it was considered that the public interface to eager loading is `config.eager_load`.

This patch is an attempt to restore that functionality, and it presents the caveats that motivated the change in Rails 6.

1. You tell the user the config endpoint can be officially overridden, so they really have no control. How can you configure something whose validity is uncertain?
3. If this method becomes public API, we need to define when can it be called.
2. 3rd party code that relies on the `:before_eager_load` initializer hook does not get their code executed if `config.eager_load` is false, because a manual call is not part of the initialization process.
4. In `classic` mode, if `config.eager_load` is false Rails unhooks AS::Dependencies after eager loading in the `:disable_dependency_loading` initializer, manual eager loading would not do that.
4. Let me also note that, since Rails 6 in `zeitwerk` mode allows applications to have autoloaded paths that are not eager loaded (because Zeitwerk is thread-safe autoloading), eager loading does not generally solve the problem of being aware of all classes and modules present in the application, which is a use case for manual eager loads.

rafaelfranca commented:
👍 for me. I think it matches my expectation on how eager load an application programatically.

fxn commented:
My concern is that the contract both for users and for 3rd party code is blurred.

Let me show you a concrete example:

```ruby
# Trigger i18n config before any eager loading has happened
# so it's ready if any classes require it when eager loaded.
config.before_eager_load do |app|
  I18n::Railtie.initialize_i18n(app)
end
```

This is from the `i18n` railtie. Since the _assumption_ is that eager loading is related to the initialization process, which in turn depends on `config.eager_load`, existing code may rely on that assumption like that gem does. In the case of that railtie, there is an `after_initialize` hook that fortunately does the same, but there is no requirement to do that.

By restricting the public API of eager loading to `config.eager_load` and the boot process, the interface is clear and well-defined for everybody. If we allow manual eager loading, I see broken windows.

Do we have use cases for manually eager load outside Rake tasks?

rafaelfranca commented:
> Do we have use cases for manually eager load outside Rake tasks?

Yes. In the test suite as well. Some tests needs to load all classes so we eager load the constants only on those tests.

I'd not care about `after_initializers` blocks since they are dealing with initialization, not eager loading, but should not we execute the `before_eager_load` blocks as well when calling `eager_load!`?

pedrofurtado commented:
Glad to see this PR merged asap 🎉 😄

fxn commented:
@pedrofurtado this is a draft, needs work.

fxn committed:
restore ability to eager load manually

fxn commented:
Retaking this discussion, I have updated the WIP patch.

I have moved more stuff into the application method, so the initializer reads now

```ruby
initializer :eager_load! do |app|
  app.eager_load! if config.eager_load
end
```

which looks good.

If client code is able to eager load manually, then we need also to ensure `eager_load!` is idempotent and thread-safe. An internal flag and a mutex take care of that.

However, I believe we are not done yet, and indeed not convinced this can be done. If you look at the [finisher](https://github.com/rails/rails/blob/667b150c20feade7138a9c0a2b0886e32e830801/railties/lib/rails/application/finisher.rb) and search for "eager", there is more logic associated.

That was my point removing this method: Eager loading is a concept that belongs to the initialization process, the application method implemented only one part of it, and so could not be the interface as things are currently defined. The public interface is `config.eage_load`.

rafaelfranca commented:
All logic I see on finisher related to eager_load is associated to the config but not to the behavior, with exception of the routes loading. I'm not sure it is worth to eager load the routes with the application.

What we need in our application is a way to eager load the constants with a single method, in the same way it is done when the application is initialized. We don't need to eager load the entire application, so what we have right now is enough to me.

fxn commented:
I think we are entering a rabbit hole here and rethinking this design needs time.

I have restored the ability of eager loading application code emulating what it was done before in https://github.com/rails/rails/commit/3285f3df7bc3a7a978428044b5458f1eed30b1a9. As the commit message explains, it is a minimal regression patch.

Can you folks please check that one out?

rafaelfranca commented:
@casperisfine @Edouard-chin can you check if we can remove our custom code and use what https://github.com/rails/rails/commit/3285f3df7bc3a7a978428044b5458f1eed30b1a9 provided?

Edouard-chin commented:
Let me check now

Edouard-chin commented on railties/lib/rails/application.rb:
```diff
@@ -580,6 +584,26 @@ def validate_secret_key_base(secret_key_base)
+
+        super unless Rails.autoloaders.zeitwerk_enabled?
+        config.eager_load_namespaces.each do |namespace|
+          namespace.eager_load! unless namespace == self
```
```suggestion
          namespace.eager_load! unless namespace == self.class
```

This breaks otherwise with a `#<NoMethodError: protected method `eager_load!' called for #<Shopify::Application:0x00007fd188a72658>>`

fxn replied:
Yes, this was untested, written in the editor to be able to discuss with actual code.

Edouard-chin commented:
works fine after the suggestion I made. Let me trigger a full CI just in case

fxn commented:
@Edouard-chin please note that the patch pushed to `6-0-stable` (linked above) is different. I felt we were entering a rabbit hole trying to put limits to something fundamentally ill-defined for 6.0.3. So, the regression patch I pushed is minimal.

Edouard-chin commented:
Ah my bad didn't see 3285f3d is a different commit, let me check

Edouard-chin commented:
I don't see any issue with https://github.com/rails/rails/commit/3285f3df7bc3a7a978428044b5458f1eed30b1a9 in our application. Our custom code is closer to this commit (from this PR) but I understand why you pushed something minimal. Anyway it works correctly, thanks !

fxn commented:
@Edouard-chin perfect, thanks for the prompt feedback :).

fxn commented:
I believe 3285f3df7bc3a7a978428044b5458f1eed30b1a9 is good as a compromise, have registered in the CHANGELOG and have cherry-picked it to `master`.

ankane commented:
Thanks @fxn!
